[https://nvbugs/5302040][feat] Add whisper support (Bert Attention on SM100 and GPTAttention for cross attention on SM100)

wu6u3tw opened the pull request:
# [nvbugs/5302040] feat. Add whisper support (Bert Attention on SM100 and GPTAttention for cross attention on SM100)

## Description
- Add bert attention plugin for Blackwell: 
    implement with fmhaDispatcher
  
- Add same seq_len causal mask support for gpt attention on cross attention:
    fix mask type and fix mask plugin layer





Please explain the issue and the solution in short.

## Test Coverage

<!--
Please list clearly what are the relevant test(s) that can safeguard the changes in the PR. This helps us to ensure we have sufficient test coverage for the PR.
-->

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Bug Fixes**
  * Improved GPU compatibility for BERT/GPT attention across architectures.
  * Limited use of custom/packed attention masks to older GPUs (SM < 100), preventing issues on newer devices.
  * Removed an unnecessary hardware guard, enabling the attention plugin on additional GPUs.

* **Refactor**
  * Internal attention kernel dispatch and initialization updated for more robust execution and support checks.
  * No public API changes; behavior remains the same for supported configurations.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

wu6u3tw committed:
rebased with whisper

Signed-off-by: tinyinl <tinyinl@nvidia.com>

nvzhihanj commented:
@PerkzZheng @kaiyux can you review or suggest the right PIC to review this PR?
This is to enable Whisper submission for MLPerf 5.1 B200 - adding support for SM100

tensorrt-cicd commented:
[PR_Github #10072](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/10072/) [ run ] triggered by Bot

nvzhihanj commented:
@wu6u3tw please run precommit hook locally and call bot run in the comment section

wu6u3tw committed:
Signed-off-by: tinyinl <tinyinl@nvidia.com>

	modified:   cpp/tensorrt_llm/common/attentionOp.cpp
	modified:   cpp/tensorrt_llm/plugins/bertAttentionPlugin/bertAttentionPlugin.cpp
	modified:   tensorrt_llm/functional.py
	modified:   tensorrt_llm/layers/attention.py

Signed-off-by: tinyinl <tinyinl@nvidia.com>

wu6u3tw commented:
/bot run

nvzhihanj commented:
/bot run

tensorrt-cicd commented:
[PR_Github #10192](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/10192/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #10192](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/10192/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #7524](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/7524/) completed with status: 'FAILURE'

PerkzZheng reviewed (commented):
Other changes LGTM. BTW, are there any tests added yet to guard this (Whisper) ?

wu6u3tw commented on cpp/tensorrt_llm/plugins/bertAttentionPlugin/bertAttentionPlugin.cpp:
```diff
@@ -520,7 +520,14 @@ int BertAttentionPlugin::enqueueImpl(nvinfer1::PluginTensorDesc const* inputDesc
                 cudaMemcpyAsync(fmhaParams.tileCounterPtr, fmha_scheduler_counter_h, sizeof(uint32_t),
                     cudaMemcpyHostToDevice, stream);
-                mFMHARunner->run(fmhaParams);
+                if (tensorrt_llm::common::getSMVersion() >= 100 && tensorrt_llm::common::getSMVersion() < 120)
```
Let me try and test it locally.

wu6u3tw committed:
dispatcher for h200

Signed-off-by: tinyinl <tinyinl@nvidia.com>

wu6u3tw commented:
/bot run

tensorrt-cicd commented:
[PR_Github #10698](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/10698/) [ run ] triggered by Bot

wu6u3tw committed:
add

Signed-off-by: tinyinl <tinyinl@nvidia.com>

wu6u3tw committed:
add

Signed-off-by: tinyinl <tinyinl@nvidia.com>

clean

Signed-off-by: tinyinl <tinyinl@nvidia.com>

clean

Signed-off-by: tinyinl <tinyinl@nvidia.com>

clean

Signed-off-by: tinyinl <tinyinl@nvidia.com>

tensorrt-cicd commented:
[PR_Github #10728](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/10728/) [ run ] triggered by Bot

nvzhihanj commented:
/bot run

tensorrt-cicd commented:
[PR_Github #11097](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11097/) [ run ] triggered by Bot

nvzhihanj commented on tensorrt_llm/functional.py:
```diff
@@ -5804,7 +5805,7 @@ def gpt_attention(
         # useFullCustomMask
         plug_inputs += [attention_mask]
-    if attention_packed_mask is not None:
+    if attention_packed_mask is not None and get_sm_version() < 100:
```
Same here, I am guessing these 2 are causing the CI issues.

wu6u3tw commented:
/bot run

tensorrt-cicd commented:
[PR_Github #12009](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/12009/) [ run ] triggered by Bot

wu6u3tw commented:
/bot run

tensorrt-cicd commented:
[PR_Github #12019](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/12019/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #12019](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/12019/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #8925](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/8925/) completed with status: 'FAILURE'

wu6u3tw commented:
/bot run --disable-fail-fast

nvzhihanj commented:
/bot run --disable-fail-fast

wu6u3tw commented on cpp/tensorrt_llm/plugins/bertAttentionPlugin/bertAttentionPlugin.h:
```diff
@@ -115,6 +116,7 @@ class BertAttentionPlugin : public BasePlugin
 
     // The default copy constructor will leave them as nullptr. clone() shall initialize it.
     UniqPtrWNullCopy<tensorrt_llm::kernels::FusedMHARunnerV2> mFMHARunner;
+    UniqPtrWNullCopy<tensorrt_llm::kernels::FmhaDispatcher> mFmhaDispatcher;
```
cleaned

tensorrt-cicd commented:
[PR_Github #12182](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/12182/) [ run ] triggered by Bot

wu6u3tw commented:
/bot run --disable-fail-fast

symphonylyh commented on cpp/tensorrt_llm/plugins/bertAttentionPlugin/bertAttentionPlugin.cpp:
```diff
@@ -520,7 +520,7 @@ int BertAttentionPlugin::enqueueImpl(nvinfer1::PluginTensorDesc const* inputDesc
                 cudaMemcpyAsync(fmhaParams.tileCounterPtr, fmha_scheduler_counter_h, sizeof(uint32_t),
                     cudaMemcpyHostToDevice, stream);
-                mFMHARunner->run(fmhaParams);
+                mFmhaDispatcher->run(fmhaParams);
```
Just to verify if I understand it correctly @PerkzZheng : the new FMHA dispatcher encapsulates the old FMHARunner & trtllmGenRunner, so it can run both fmha_v2 (<= sm90) and trtllm-gen (>=sm100)

PerkzZheng replied:
yes.

symphonylyh commented on cpp/tensorrt_llm/plugins/bertAttentionPlugin/bertAttentionPlugin.cpp:
```diff
@@ -948,10 +959,15 @@ int BertAttentionPlugin::initialize() noexcept
-
-        // Fall back to unfused MHA kernels if not supported.
-        mEnableContextFMHA = mFMHARunner->isFmhaSupported();
+        // // The KV input data type. The default is same as dataType.
```
yes, using fmhadispatcher is correct, but the code comment "Load kernels from the pre-compiled cubins for blackwell" is not? on Hopper it loads hopper cubins, on Blackwell it loads blackwell cubins

wu6u3tw commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #12412](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/12412/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #12412](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/12412/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #9228](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/9228/) completed with status: 'FAILURE'

wu6u3tw commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #13731](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13731/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #10318](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/10318/) completed with status: 'FAILURE'

wu6u3tw commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #14040](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14040/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #14327](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14327/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #10825](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/10825/) completed with status: 'SUCCESS'

tensorrt-cicd commented:
[PR_Github #14640](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14640/) [ reuse-pipeline ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #14640](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14640/) [ reuse-pipeline ] completed with state `SUCCESS`
Reusing [PR_Github #14327](https://prod.blsm.nvidia.com/sw-tensorrt-top-1/job/LLM/job/helpers/job/PR_Github/14327/) for commit `b23b323`

PerkzZheng commented on cpp/tensorrt_llm/plugins/bertAttentionPlugin/bertAttentionPlugin.cpp:
```diff
@@ -947,11 +958,15 @@ int BertAttentionPlugin::initialize() noexcept
-        mFMHARunner.reset(new FusedMHARunnerV2(fmhaParams));
+        // The KV input data type. The default is same as dataType.
+        fmhaParams.dataTypeKv = data_type;
+        fmhaParams.forceFp32Acc = false;
```
should this be equal to `mFMHAForceFP32Acc` ?

PerkzZheng reviewed (approved):
Others LGTM.

tensorrt-cicd commented:
[PR_Github #15013](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15013/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #11340](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/11340/) completed with status: 'FAILURE'

wu6u3tw commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #15018](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15018/) [ run ] triggered by Bot

wu6u3tw commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #15025](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15025/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #15018](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15018/) [ run ] completed with state `ABORTED`

wu6u3tw commented:
/bot run --disable-fail-fast

wu6u3tw commented:
/bot run

tensorrt-cicd commented:
[PR_Github #15031](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15031/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #15029](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15029/) [ run ] completed with state `ABORTED`

tensorrt-cicd commented:
[PR_Github #15031](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15031/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #11353](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/11353/) completed with status: 'SUCCESS'
Pipeline passed with automatic retried tests. Check the [rerun report](https://urm.nvidia.com/artifactory/sw-tensorrt-generic/llm-artifacts//LLM/main/L0_MergeRequest_PR/11353/test-results/rerun_report.html) for details.
